### PR TITLE
Streamline native injection w.r.t cloning

### LINKED ISSUE #1994: Duplicate Tests being added to XmlTests if XmlTest is used in the parameters of annotated methods
### TestNG Version
v7

### Expected behavior
Current XmlTest is returned

### Actual behavior
Duplicate test is added to XmlTest

### Is the issue reproductible on runner?

- [] Shell
- [] Maven
- [] Eclipse

### Test case sample
For example, use the following `@BeforeTest`

```java
@BeforeTest(alwaysRun = true)
@Parameters({ "browser"})
 public void beforeTest(@Optional XmlTest xmlTest, @Optional String browser) {
      System.setProperty("webdriver.chrome.driver","/path/to/chromedriver");
      DesiredCapabilities c=DesiredCapabilities.chrome();
      ChromeOptions options = new ChromeOptions();
      options.addArguments("options");
}
```

Run any sample suite with around 5 test cases
Now after the tests are executed, check the total number of tests using debugging mode in m_tests. It shows 8 tests which include duplicate tests being added which contain empty xmlClasses
image

These duplicate/ empty test cases are also causing random NullPointer Exceptions at the end of the tests. These NPEs are however not reproducible all the time.

Attached is a small zip file with a sample maven project. You can reproduce the duplicate tests issue by just running the `testng.xml` file from eclipse and checking the `m_tests` at the `AfterSuite` methods by using `ITestContext` like the below:

```java
@AfterSuite
public void afterSuite(ITestContext context) {
	  System.out.println("afterSuite");
}
```

Please let me know if any other info is needed 

[DuplicateTests.zip](https://github.com/cbeust/testng/files/2717795/DuplicateTests.zip)

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #1994

Currently as part of supporting native injection
TestNG resorts to invoking the “clone()” method 
on the parameter being natively injected.

This becomes a problem because XmlTest is the only
internal TestNG object that implements the Cloneable
interface wherein it resorts to something like a deep
copy and adds the cloned instance to the XmlSuite.

This causes a lot of phony XmlTest objects to be 
added up to the suite and causes issues as detailed
in the bug.

Fixed this by introducing a marker interface to 
indicate to TestNG to skip cloning and use the object
as is. Annotated “XmlTest” with this annotation 
so that we dont clone xmltest but instead use it 
as is.

Yes this will expose the XmlTest object, but its 
already exposed to the end user via various different
means. So I guess we are ok here.

Fixes #1994 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

## Summary by CodeRabbit

- **New Features**
	- Introduced a method to allow overriding error messages in assertions.
	- Added `@SkipCloning` annotation to prevent cloning of objects for dependency injection.
	- Enhanced object copying functionality with a new copy constructor and deprecated the old cloning method in `XmlTest`.
	- Added a test to ensure native injection does not resort to cloning.
- **Bug Fixes**
	- Fixed duplication issues with `XmlTest` objects used as parameters.
	- Resolved execution problems with `@AfterGroups` when group members fail or are skipped.
	- Ensured `@BeforeGroups` is invoked only when groups are explicitly specified.
- **Documentation**
	- Deprecated methods and usage instructions updated across various components.

### CODE REVIEW

**Comment by krmahadevan** [comment]
@juherr - As decided in https://github.com/cbeust/testng/pull/1996 I have added a copy constructor, deprecated the `clone()` method and also chose to retain the newly introduced annotation (I feel that if its evangelized properly it would help TestNG to skip costly clone methods in the user defined object that is being injected as a parameter).

File: src/main/java/org/testng/annotations/SkipCloning.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package org.testng.annotations;
```
[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/main/java/org/testng/annotations/SkipCloning.java]: @juherr - Here's my rationale behind leaving this exposed ( I also would like to evangelize about this lesser known behavior of TestNG )

Currently when it comes to parameter injection and associating that with the `TestResult` object, we resort to `cloning()` if the user's parameter object class implements the `Cloneable` interface. So there could be a chance wherein the parameters that the user is altering in their tests is perhaps not reflected in their listeners or in their reports because we cloned them.

I think we should enrich our documentation to first call out this behavior explicitly ( a most common use case is a data driven test that is run via a `@DataProvider` annotation) and via this newly introduced annotation, still let the user control whether they want or do not want their test objects to be cloned by TestNG.

**Comment by krmahadevan** [comment]
Rebased off of `master` and resolved merge conflicts.

File: src/main/java/org/testng/annotations/SkipCloning.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (juherr) [reply src/main/java/org/testng/annotations/SkipCloning.java]: I disagree because you want to make public something which wasn't asked.
Then, the clone solution was just a quick workaround for a design issue when the same object is used in reports. But the clone solution doesn't cover all the need.

That's why I have some doubts to publish a workaround of a workaround before be really sure how we need to fix the root issue.

[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/main/java/org/testng/annotations/SkipCloning.java]: > I disagree because you want to make public something which wasn't asked.

The changeset that introduced setting cloned variants of parameters was to fix https://github.com/cbeust/testng/issues/447 It was done by you via commit https://github.com/cbeust/testng/commit/78001fd0ed81970784b9d17b887998064fff0c03

Via this commit we ended up implicitly cloning a parameter (if it was clone able) but never made this information public. IMO we should have done this, because an end-user should know if the real parameter that is being passed to the test method is being used or if its a cloned one. 

We haven't heard of anyone mention about this problem only because usually our users dont implement the `Cloneable` interface to the parameter that they are passing to a data provider.

The moment someone does that, this implicit behavior will become evident. 

So now it looks like the root cause was us introducing the Cloneable parameter.  
What if the user provided parameter was clone able but the user didn't want it to be clone able ? How would they be able to support that. With collections it may not be possible because a user cannot add that annotation to a collection class (I think this also should be possible if we just enrich `@SkipCloning` annotation such that it can be added to a parameter just as how a user adds `@Optional`), but at least with a user's test project defined objects it should be something that they should be able to do.

This was my thinking.

[the remark this answers is not in the document]
Comment (juherr) [reply src/main/java/org/testng/annotations/SkipCloning.java]: Yes, I understand your point of view. But mine is the #447 was just a not complete workaround of the root issue.
I agree to release another workaround to fix the first workaround but I'm not feeling it good to make it public because we will have to manage deprecation.
Instead, I prefer to make it private and to find a good solution for the root issue.
Furthermore, I'm not sure my old commit will still work on with Java9 modules due to` clone.setAccessible(true)`.

For the current PR:
* private: OK and waiting for a better fix for #447
* public: not yet ok :)

[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/main/java/org/testng/annotations/SkipCloning.java]: ping @juherr

[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/main/java/org/testng/annotations/SkipCloning.java]: Without giving the control of deciding which parameter to be cloned and which not to, how do you propose the problem will be fixed ? Can you please help elaborate that ? 

@cbeust what are you thoughts on this ?

**Comment by juherr** [comment]
@coderabbitai review

@krmahadevan Currently testing a new. tool

**Comment by juherr** [comment]
@coderabbitai review

**Comment by juherr** [comment]
@coderabbitai review

**Comment by juherr** [comment]
Closing this PR as superseded by #3401.

#3401 fixes #1994 by keeping injected `XmlTest` instances by reference when snapshotting `ITestResult` parameters, with regression coverage for the original scenario. This avoids introducing the public `@SkipCloning` API proposed here.
